docs/protocol.md

```diff
@@ -0,0 +1,364 @@
+# HTTP protocol for RAG chat apps
+
+We are standardizing on a common HTTP protocol across RAG chat app solutions and tools,
+to make them more compatible with each other. By agreeing on a common protocol, developers
+can swap different components, like using the Python backend from this repository
+with a Web Components frontend from [azure-search-openai-javascript](https://github.com/Azure-Samples/azure-search-openai-javascript).
+
+This protocol is inspired by the [OpenAI ChatCompletion API](https://platform.openai.com/docs/guides/text-generation/chat-completions-api),
+but contains additional fields required for a RAG chat app.
+
+## HTTP requests to RAG chat app endpoints
+
+An HTTP request to `/chat` or `/ask` can contain these properties, in JSON:
+
+* `"messages"`: A list of messages, each containing "content" and "role", where "role" may be "system", "assistant", or "user"
+* `"stream"`: A boolean indicating whether the response should be streamed or not.
+* `"context"`: An object containing any additional context about the request. This app only sets the "overrides" property, which is an object with additional settings for the request, like how many search results to return.
+* `"session_state"`: An object containing the "memory" for the chat app, such as a user ID. Not currently used by this app.
+
+Here's an example JSON request:
+
+```json
+{
+    "messages": [
+        {
+            "content": "What is included in my Northwind Health Plus plan that is not in standard?",
+            "role": "user"
+        }
+    ],
+    "stream": false,
+    "context": {
+        "overrides": {
+            "top": 3,
+            "retrieval_mode": "text",
+            "semantic_ranker": false,
+            "semantic_captions": false,
+            "suggest_followup_questions": false,
+            "use_oid_security_filter": false,
+            "use_groups_security_filter": false,
+            "vector_fields": ["embedding"],
+            "use_gpt4v": false,
+            "gpt4v_input": "textAndImages"
+        }
+    },
+    "session_state": null
```

Comment:
Is there anything interesting about `session_state` that is specific to OpenAI/RAG?

Reply:
I think the idea was to utilize it for conversation history and user authentication, but we don't have conversation history and we use headers for user ID, so it's null for us. We do honor it though, so we pass it back and forth. There's more about it in Dan's original doc, if you can view this: https://github.com/Azure/azureml_run_specification/pull/1066/files